cmd/sonobuoy/app/images.go

```diff
@@ -101,3 +157,195 @@ func getImages(cmd *cobra.Command, args []string) {
+	switch imagesflags.plugin {
+	case "e2e":
+
+		cfg, err := imagesflags.kubeconfig.Get()
+		if err != nil {
+			errlog.LogError(errors.Wrap(err, "couldn't get REST client"))
+			os.Exit(1)
+		}
+
+		sbc, err := getSonobuoyClient(cfg)
+		if err != nil {
+			errlog.LogError(errors.Wrap(err, "could not create sonobuoy client"))
+			os.Exit(1)
+		}
+
+		version, err := sbc.Version()
+		if err != nil {
+			errlog.LogError(errors.Wrap(err, "couldn't get Sonobuoy client"))
+			os.Exit(1)
+		}
+
+		upstreamImages, err := image.GetImages("", version)
+		if err != nil {
+			errlog.LogError(errors.Wrap(err, "couldn't init upstream registry list"))
+			os.Exit(1)
+		}
+
+		images := []string{}
+		for _, v := range upstreamImages {
+			images = append(images, v.GetE2EImage())
+		}
+		err = docker.Save(images, image.GetTarFileName(version))
+		if err != nil {
+			errlog.LogError(errors.Wrap(err, "couldn't save images to tar"))
+			os.Exit(1)
+		}
+
+	default:
+		errlog.LogError(errors.Errorf("Unsupported plugin: %v", imagesflags.plugin))
+		os.Exit(1)
+	}
+}
+
+func pushImages(cmd *cobra.Command, args []string) {
+
+	switch imagesflags.plugin {
+	case "e2e":
+
+		if len(imagesflags.e2eRegistryConfig) > 0 {
+			// Check if the e2e file exists
+			if _, err := os.Stat(imagesflags.e2eRegistryConfig); os.IsNotExist(err) {
+				errlog.LogError(errors.Errorf("file does not exist or cannot be opened: %v", imagesflags.e2eRegistryConfig))
+				os.Exit(1)
+			}
+		}
+
+		// Check if the e2e file exists
+		if _, err := os.Stat(imagesflags.e2eRegistryConfig); os.IsNotExist(err) {
+			errlog.LogError(errors.Errorf("file does not exist or cannot be opened: %v", imagesflags.e2eRegistryConfig))
+			os.Exit(1)
+		}
+
+		cfg, err := imagesflags.kubeconfig.Get()
+		if err != nil {
+			errlog.LogError(errors.Wrap(err, "couldn't get REST client"))
+			os.Exit(1)
+		}
+
+		sbc, err := getSonobuoyClient(cfg)
+		if err != nil {
+			errlog.LogError(errors.Wrap(err, "could not create sonobuoy client"))
+			os.Exit(1)
+		}
+
+		version, err := sbc.Version()
+		if err != nil {
+			errlog.LogError(errors.Wrap(err, "couldn't get Sonobuoy client"))
+			os.Exit(1)
+		}
+
+		upstreamImages, err := image.GetImages("", version)
+		if err != nil {
+			errlog.LogError(errors.Wrap(err, "couldn't init upstream registry list"))
+			os.Exit(1)
+		}
+
+		privateImages, err := image.GetImages(imagesflags.e2eRegistryConfig, version)
+		if err != nil {
+			errlog.LogError(errors.Wrap(err, "couldn't init upstream registry list"))
+			os.Exit(1)
+		}
+
+		for k, v := range upstreamImages {
+			privateImg := privateImages[k]
+			err = docker.Tag(v.GetE2EImage(), privateImg.GetE2EImage(), numDockerRetries)
+			if err != nil {
+				errlog.LogError(errors.Wrapf(err, "couldn't tag image: %v", v.GetE2EImage()))
+			}
+
+			err = docker.Push(privateImg.GetE2EImage(), numDockerRetries)
+			if err != nil {
+				errlog.LogError(errors.Wrapf(err, "couldn't push image: %v", v.GetE2EImage()))
```

Comment:
For these errors we may want to look at how we log them. When shelling out to other commands you get a certain type of error which doesnt print useful info, just the exit status:
```
ERRO[0005] couldn't push image: gcr.io/google-samples/gb-redisslave:v3: exit status 1
```
Stderr has the useful info but we may inspect the error and see if there is another piece of useful data there instead.

Reply:
I think the solution to this is to run using `exec.RunLoggingOutputOnFailed`. I got the following output which is much more clear:

```
INFO[0026] Trying again to push image: docker.io/schnake/volume/nfs:1.0 ...  error="exit status 1"
ERRO[0027] failed with:
ERRO[0027] The push refers to repository [docker.io/schnake/volume/nfs]
ERRO[0027] 37970ad80d97: Preparing
ERRO[0027] 37970ad80d97: Preparing
ERRO[0027] 1448ce2d0d0c: Preparing
ERRO[0027] 2cedbfe57c8f: Preparing
ERRO[0027] 4d03698a8437: Preparing
ERRO[0027] 1d31b5806ba4: Preparing
ERRO[0027] denied: requested access to the resource is denied
```